### ISSUE TITLE
CascadeClassifier::detectMultiScale levelWeights & rejectLevels

### ISSUE THREAD

**Issue by sam-n-johnston** [opencv/opencv#8016]
##### System information (version)
- OpenCV => 3.2.0
- Operating System / Platform => Ubuntu 14.04
- Compiler => gcc 4.8.4

##### Detailed description

I'm having a problem trying to implement a code that determines if an image is a face. I created a minimal [repo](https://github.com/sam-n-johnston/cascade-weight-test/tree/master) to test this out. There are six face, from a good relatively fontal face to a decentered face. The output I get is strange: 
```
Detection ./cropped_face1.png [109 x 109 from (2, 14)] with weight 108.512 and level: 22 and size: 1
Detection ./cropped_face2.png [111 x 111 from (9, 9)] with weight 108.411 and level: 22 and size: 1
Detection ./cropped_face3.png [109 x 109 from (2, 14)] with weight 108.512 and level: 22 and size: 1
```
The last 3 faces do not even get a single output, even if face#4 is reasonably nice:
Face#3:![cropped_face3](https://cloud.githubusercontent.com/assets/17952091/21995309/2e7e0d12-dbf2-11e6-8768-2b7305d83a2f.png)
Face#4:![cropped_face4](https://cloud.githubusercontent.com/assets/17952091/21995009/9b530d22-dbf0-11e6-88f4-591d9b1de1f6.png)

The level always seems to be 22 (on my very little sample, sorry). Stage 22 seems to be the final stage of the classifier based on the xml file of the classifier. I am also unclear on to what the weights actually represent with such strange number. I didn't find anything in the documentation to help me. I started going through the source code, but I didn't find anything to help me yet. 

I do not quite understand why face#3 gets to the final stage of the classifier, with ~108 of weight, and the face#4 has no output, it didn't even get to stage 1? 

Am I doing something wrong? Is something wrong with the function? Can there be an explanation of the usage and what the weight is? 

##### Steps to reproduce

See [repo](https://github.com/sam-n-johnston/cascade-weight-test/tree/master)

**Comment by StevenPuttemans** [comment]
Some answers to your questions

> The level always seems to be 22 (on my very little sample, sorry). 

Yes that is true. The whole idea behind returning the level of rejection was so broken, that we decided for now to output only samples going through your whole cascade. This makes the most sense, since earlier rejected frames should not be allowed to be visualised as detections. Why else train a 22 stage classifier?

>  I am also unclear on to what the weights actually represent with such strange number

You should get yourself a copy of OpenCV 3 Blueprints, Chapter 5, which discusses all these parameters in the finest details. That being said, the weights are the decision level of your final stage classifier.

IN GENERAL

It seems that you are using a detector for classification, which it is not designed for. My guess, your faces are to close to the border, which makes the detector fail. Again this is designed to locate face areas in an image with background information.

**Comment by sam-n-johnston** [comment]
> Why else train a 22 stage classifier?

To get a % confidence of classification for an image. I had to "hack" OpenCV 2.4 to get this functionality, and it was working nicely! I'm not the only one with this need as I've seen at multiple locations people indicating how to do this. I was basing myself on this PR and it's response: [#7728](https://github.com/opencv/opencv/pull/7728#issuecomment-263830190). Reading this, I thought I would get what I needed with this function: a confidence score for an image patch based on rejected level + score at that level. 

> The whole idea behind returning the level of rejection was so broken, that we decided for now to output only samples going through your whole cascade.

So, basically this overloaded function doesn't have much use now? 

Ok, it makes sense that the faces are too close to the border and that's why they aren't classified or detected. 

Do you have a recommendation on how I can achieve a classification score from a detector or OpenCV?

**Comment by arpit1997** [comment]
> If you want a pure classification, why not go for SVM, calculate some HOG features, or some LBP feature histogram and then do a simple binary classifier.
This is what exactly the docs in [opencv](https://github.com/opencv/opencv). even the face recognizers use this method 
:smiley: :smile_cat: :small_blue_diamond:

**Comment by sturkmen72** [comment]
`cv::CascadeClassifier` is almost deprecated. i think this issue can be closed down.

**Comment by StevenPuttemans** [comment]
> 
> 
> `cv::CascadeClassifier` is almost deprecated. i think this issue can be closed down.

I am sorry @sturkmen72 but I do not agree that cascade classifiers are deprecated. Yes DL detectors work as smooth and fast and are more acurate, but these still have their applications in embedded setups :)

**Comment by CognitiveClouds-Prasad** [comment]
> `cv::CascadeClassifier` is almost deprecated. i think this issue can be closed down.
Citation needed.
